### PR TITLE
[Performance] Allow DataLoaders to return indices of the seed array

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
As requested in #2461 .

For semi-supervised learning with very few labels, creating a label tensor with the same size as the number of nodes wastes memory.  This change introduces the optional argument `return_indices` that allows the DataLoader to additionally return the indices of seed nodes/edges sampled in the minibatch.

### Example 1

```python
g = dgl.rand_graph(100, 1000)
sampler = dgl.dataloading.MultiLayerNeighborSampler([4])
dl = dgl.dataloading.NodeDataLoader(
    g, torch.arange(0, 100, 5), sampler, batch_size=5, return_indices=True, shuffle=True)
next(iter(dl))
```
Output:
```
(tensor([15, 20,  0, 95, 35,  8, 78, 99, 73, 97,  6, 40,  2, 14, 92, 54, 16, 23,
         30, 71, 50, 94, 93]),
 tensor([15, 20,  0, 95, 35]),
 tensor([ 3,  4,  0, 19,  7]),
 [Block(num_src_nodes=23, num_dst_nodes=5, num_edges=20)])
```
### Example 2
```python
dl = dgl.dataloading.EdgeDataLoader(
    g, torch.arange(0, 1000, 5), sampler, batch_size=5, return_indices=True, shuffle=True,
    negative_sampler=dgl.dataloading.negative_sampler.Uniform(2))
result = next(iter(dl))
result, result[1].edata[dgl.EID]
```
Output:
```
((tensor([ 3, 86, 57, 21, 31, 90,  7, 92, 29, 13, 95, 99, 84, 54,  2, 98, 79, 60,
           8, 69, 41, 24,  5, 33, 16, 65, 63, 45, 71, 39, 97, 82, 42, 35, 73, 18,
          49, 44, 40, 74, 37, 23, 30, 75, 36,  6, 20, 51, 15, 47, 78]),
  Graph(num_nodes=14, num_edges=5,
        ndata_schemes={'_ID': Scheme(shape=(), dtype=torch.int64)}
        edata_schemes={'_ID': Scheme(shape=(), dtype=torch.int64)}),
  Graph(num_nodes=14, num_edges=10,
        ndata_schemes={'_ID': Scheme(shape=(), dtype=torch.int64)}
        edata_schemes={}),
  tensor([ 78,  15,  16,  25, 181]),
  [Block(num_src_nodes=51, num_dst_nodes=14, num_edges=56)]),
 tensor([390,  75,  80, 125, 905]))
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

### CODE REVIEW

File: python/dgl/dataloading/dataloader.py [hunk 1]
```diff
@@ -351,26 +378,34 @@ def collate(self, items):
 
             If the original graph has multiple node types, return a dictionary of
             node type names and node ID tensors.  Otherwise, return a single tensor.
+        items : Tensor or dict[ntype, Tensor], optional
```
Comment (classicsong) [inline python/dgl/dataloading/dataloader.py:381]: The name items is misleading. This should be the indices.

Comment (classicsong) [reply python/dgl/dataloading/dataloader.py:381]: Further, in Parameters, the items can also be indices.

File: python/dgl/dataloading/dataloader.py [hunk 2]
```diff
@@ -450,6 +485,9 @@ class EdgeCollator(Collator):
 
         A set of builtin negative samplers are provided in
         :ref:`the negative sampling module <api-dataloading-negative-sampling>`.
+    return_eids : bool, default False
+        Whether to additionally return the indices of the input ``nids`` array sampled in the
```
Comment (classicsong) [inline python/dgl/dataloading/dataloader.py:489]: input ``eids``

File: python/dgl/dataloading/dataloader.py [hunk 3]
```diff
@@ -575,38 +620,53 @@ def _collate(self, items):
...
-            items = utils.prepare_tensor_dict(self.g_sampling, items, 'items')
+            items = _prepare_tensor_dict(self.g_sampling, items, 'items', self._is_distributed)
         else:
-            items = utils.prepare_tensor(self.g_sampling, items, 'items')
+            items = _prepare_tensor(self.g_sampling, items, 'items', self._is_distributed)
 
-        pair_graph = self.g.edge_subgraph(items, preserve_nodes=True)
+        if isinstance(items, dict):
+            sample_items = {k: F.gather_row(self.eids[k], v) for k, v in items.items()} \
+                           if self.return_indices else items
```
Comment (classicsong) [inline python/dgl/dataloading/dataloader.py:662]: One question about negative sampling is that can we sample pos edges belong to a specific positive edge set and sample negative edges from another edge set?

Comment (BarclayII) [reply python/dgl/dataloading/dataloader.py:662]: Can you do it by writing your own negative sampler?  Do you also need indices?

File: python/dgl/dataloading/dataloader.py (hunk 1 above)
Comment (BarclayII) [reply python/dgl/dataloading/dataloader.py:381]: I'm not sure if the argument name should be indices.  It can also represent the actual node IDs if `return_indices` is False.

Comment (classicsong) [reply python/dgl/dataloading/dataloader.py]: It is returned only when return_indices is True

File: python/dgl/dataloading/dataloader.py (hunk 3 above)
Comment (classicsong) [reply python/dgl/dataloading/dataloader.py:668]: Writing my own negative sampler always can work.

**Review by classicsong — APPROVED** [review]
Some small comments.
Overall LGTM.

File: python/dgl/dataloading/dataloader.py (hunk 1 above)
Comment (BarclayII) [reply python/dgl/dataloading/dataloader.py]: It's right below.
